src/engine/thread.js

```diff
@@ -321,14 +340,22 @@ class Thread {
         this.justReported = typeof value === 'undefined' ? null : value;
     }
 
+    /**
+     * Return a execution context for a block to use.
+     * @returns {object} the execution context
+     */
+    getExecutionContext () {
+        const frame = this.stackFrame;
+        frame.needReset = true;
```

Comment:
Why does looking at the executionContext indicate that the frame needs a reset?

Reply:
With this PR most StackFrame properties remain do not need to change from when the frame is initialized to when the frame is released into the free list. isLoop and executionContext (and needReset) are the remaining properties that may change.

We need to reset isLoop and executionContext when we reuseStackForNextBlock or popStack on that stack frame. Or that we need to reset executionContext if it is used. I thought this may be a way we capture that info. If a block doesn't get the executionContext then it cannot have used it and we can skip a bit of work resetting it.

There is an alternative, similar to what was in block utility and with previous thread code. When getExecutionContext is called it'll assign an object to executionContext if needed. Then when we reuse or release we can set executionContext to null.

Those are kind of the same and not. One assumes we need to reset after use, and the other assumes we need to create once used.